File: src/TypescriptSDK/src/models/SimulationControl.ts
import { IsString, IsOptional, Matches, IsBoolean, validate, ValidationError as TsValidationError } from 'class-validator';
import { Type, plainToClass, instanceToPlain, Expose, Transform } from 'class-transformer';
import { _OpenAPIGenBaseModel } from "./_OpenAPIGenBaseModel";

/** Used to specify which types of calculations to run. */
export class SimulationControl extends _OpenAPIGenBaseModel {
    @IsString()
    @IsOptional()
    @Matches(/^SimulationControl$/)
    @Expose({ name: "type" })
    /** type */
    type: string = "SimulationControl";
	
    @IsBoolean()
    @IsOptional()
    @Expose({ name: "do_zone_sizing" })
    /** Boolean for whether the zone sizing calculation should be run. */
    doZoneSizing: boolean = true;
	
    @IsBoolean()
    @IsOptional()
    @Expose({ name: "do_system_sizing" })
    /** Boolean for whether the system sizing calculation should be run. */
    doSystemSizing: boolean = true;
	
    @IsBoolean()
    @IsOptional()
    @Expose({ name: "do_plant_sizing" })
    /** Boolean for whether the plant sizing calculation should be run. */
    doPlantSizing: boolean = true;
	
    @IsBoolean()
    @IsOptional()
    @Expose({ name: "run_for_run_periods" })
    /** Boolean for whether the simulation should be run for the sizing periods. */
    runForRunPeriods: boolean = true;
	
    @IsBoolean()
    @IsOptional()
    @Expose({ name: "run_for_sizing_periods" })
    /** Boolean for whether the simulation should be run for the run periods. */
    runForSizingPeriods: boolean = false;
	

    constructor() {
        super();
        this.type = "SimulationControl";
        this.doZoneSizing = true;
        this.doSystemSizing = true;
        this.doPlantSizing = true;
        this.runForRunPeriods = true;
        this.runForSizingPeriods = false;
    }


    override init(_data?: any) {
        super.init(_data);
        if (_data) {
            const obj = plainToClass(SimulationControl, _data, { enableImplicitConversion: true });
            this.type = obj.type;
            this.doZoneSizing = obj.doZoneSizing;
            this.doSystemSizing = obj.doSystemSizing;
            this.doPlantSizing = obj.doPlantSizing;
            this.runForRunPeriods = obj.runForRunPeriods;
            this.runForSizingPeriods = obj.runForSizingPeriods;
        }
    }


    static override fromJS(data: any): SimulationControl {
        data = typeof data === 'object' ? data : {};

        if (Array.isArray(data)) {
            const obj:any = {};
            for (var property in data) {
                obj[property] = data[property];
            }
            data = obj;
        }
        let result = new SimulationControl();
        result.init(data);
        return result;
    }

	override toJSON(data?: any) {
        data = typeof data === 'object' ? data : {};
        data["type"] = this.type;
        data["do_zone_sizing"] = this.doZoneSizing;
        data["do_system_sizing"] = this.doSystemSizing;
        data["do_plant_sizing"] = this.doPlantSizing;
        data["run_for_run_periods"] = this.runForRunPeriods;
        data["run_for_sizing_periods"] = this.runForSizingPeriods;
        data = super.toJSON(data);
        return instanceToPlain(data);
    }

	async validate(): Promise<boolean> {
        const errors = await validate(this);
        if (errors.length > 0){
			const errorMessages = errors.map((error: TsValidationError) => Object.values(error.constraints || [error]).join(', ')).join('; ');
      		throw new Error(`Validation failed: ${errorMessages}`);
		}
        return true;
    }
}

File: src/CSharpSDK/Model/SimulationControl.cs
﻿/* 
 * HoneybeeSchema
 *
 * Contact: info@ladybug.tools
 */

extern alias LBTNewtonSoft;
//using System;
using System.Linq;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Runtime.Serialization;
using LBTNewtonSoft::Newtonsoft.Json;
using LBTNewtonSoft::Newtonsoft.Json.Converters;
using System.ComponentModel.DataAnnotations;

namespace HoneybeeSchema
{
    /// <summary>
    /// Used to specify which types of calculations to run.
    /// </summary>
    [Summary(@"Used to specify which types of calculations to run.")]
    [System.Serializable]
    [DataContract(Name = "SimulationControl")]
    public partial class SimulationControl : OpenAPIGenBaseModel, System.IEquatable<SimulationControl>
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="SimulationControl" /> class.
        /// </summary>
        [LBTNewtonSoft.Newtonsoft.Json.JsonConstructorAttribute]
        [System.Text.Json.Serialization.JsonConstructor]
        protected SimulationControl() 
        { 
            // Set readonly properties with defaultValue
            this.Type = "SimulationControl";
        }
        /// <summary>
        /// Initializes a new instance of the <see cref="SimulationControl" /> class.
        /// </summary>
        /// <param name="doZoneSizing">Boolean for whether the zone sizing calculation should be run.</param>
        /// <param name="doSystemSizing">Boolean for whether the system sizing calculation should be run.</param>
        /// <param name="doPlantSizing">Boolean for whether the plant sizing calculation should be run.</param>
        /// <param name="runForRunPeriods">Boolean for whether the simulation should be run for the sizing periods.</param>
        /// <param name="runForSizingPeriods">Boolean for whether the simulation should be run for the run periods.</param>
        public SimulationControl
        (
            bool doZoneSizing = true, bool doSystemSizing = true, bool doPlantSizing = true, bool runForRunPeriods = true, bool runForSizingPeriods = false
        ) : base()
        {
            this.DoZoneSizing = doZoneSizing;
            this.DoSystemSizing = doSystemSizing;
            this.DoPlantSizing = doPlantSizing;
            this.RunForRunPeriods = runForRunPeriods;
            this.RunForSizingPeriods = runForSizingPeriods;

            // Set readonly properties with defaultValue
            this.Type = "SimulationControl";

            // check if object is valid, only check for inherited class
            if (this.GetType() == typeof(SimulationControl))
                this.IsValid(throwException: true);
        }

	
	
        /// <summary>
        /// Boolean for whether the zone sizing calculation should be run.
        /// </summary>
        [Summary(@"Boolean for whether the zone sizing calculation should be run.")]
        [DataMember(Name = "do_zone_sizing")] // For Newtonsoft.Json
        [System.Text.Json.Serialization.JsonPropertyName("do_zone_sizing")] // For System.Text.Json
        public bool DoZoneSizing { get; set; } = true;

        /// <summary>
        /// Boolean for whether the system sizing calculation should be run.
        /// </summary>
        [Summary(@"Boolean for whether the system sizing calculation should be run.")]
        [DataMember(Name = "do_system_sizing")] // For Newtonsoft.Json
        [System.Text.Json.Serialization.JsonPropertyName("do_system_sizing")] // For System.Text.Json
        public bool DoSystemSizing { get; set; } = true;

        /// <summary>
        /// Boolean for whether the plant sizing calculation should be run.
        /// </summary>
        [Summary(@"Boolean for whether the plant sizing calculation should be run.")]
        [DataMember(Name = "do_plant_sizing")] // For Newtonsoft.Json
        [System.Text.Json.Serialization.JsonPropertyName("do_plant_sizing")] // For System.Text.Json
        public bool DoPlantSizing { get; set; } = true;

        /// <summary>
        /// Boolean for whether the simulation should be run for the sizing periods.
        /// </summary>
        [Summary(@"Boolean for whether the simulation should be run for the sizing periods.")]
        [DataMember(Name = "run_for_run_periods")] // For Newtonsoft.Json
        [System.Text.Json.Serialization.JsonPropertyName("run_for_run_periods")] // For System.Text.Json
        public bool RunForRunPeriods { get; set; } = true;

        /// <summary>
        /// Boolean for whether the simulation should be run for the run periods.
        /// </summary>
        [Summary(@"Boolean for whether the simulation should be run for the run periods.")]
        [DataMember(Name = "run_for_sizing_periods")] // For Newtonsoft.Json
        [System.Text.Json.Serialization.JsonPropertyName("run_for_sizing_periods")] // For System.Text.Json
        public bool RunForSizingPeriods { get; set; } = false;


        /// <summary>
        /// Returns the string presentation of the object
        /// </summary>
        /// <returns>String presentation of the object</returns>
        public override string ToString()
        {
            return "SimulationControl";
        }


        /// <summary>
        /// Returns the string presentation of the object
        /// </summary>
        /// <returns>String presentation of the object</returns>
        public override string ToString(bool detailed)
        {
            if (!detailed)
                return this.ToString();
            
            var sb = new StringBuilder();
            sb.Append("SimulationControl:\n");
            sb.Append("  Type: ").Append(this.Type).Append("\n");
            sb.Append("  DoZoneSizing: ").Append(this.DoZoneSizing).Append("\n");
            sb.Append("  DoSystemSizing: ").Append(this.DoSystemSizing).Append("\n");
            sb.Append("  DoPlantSizing: ").Append(this.DoPlantSizing).Append("\n");
            sb.Append("  RunForRunPeriods: ").Append(this.RunForRunPeriods).Append("\n");
            sb.Append("  RunForSizingPeriods: ").Append(this.RunForSizingPeriods).Append("\n");
            return sb.ToString();
        }


        /// <summary>
        /// Returns the object from JSON string
        /// </summary>
        /// <returns>SimulationControl object</returns>
        public static SimulationControl FromJson(string json)
        {
            var obj = JsonConvert.DeserializeObject<SimulationControl>(json, JsonSetting.AnyOfConvertSetting);
            if (obj == null)
                return null;
            return obj.Type.ToLower() == obj.GetType().Name.ToLower() && obj.IsValid(throwException: true) ? obj : null;
        }




        /// <summary>
        /// Creates a new instance with the same properties.
        /// </summary>
        /// <returns>SimulationControl object</returns>
        public virtual SimulationControl DuplicateSimulationControl()
        {
            return FromJson(this.ToJson());
        }


        /// <summary>
        /// Creates a new instance with the same properties.
        /// </summary>
        /// <returns>OpenAPIGenBaseModel</returns>
        public override OpenAPIGenBaseModel DuplicateOpenAPIGenBaseModel()
        {
            return DuplicateSimulationControl();
        }


        /// <summary>
        /// Returns true if objects are equal
        /// </summary>
        /// <param name="input">Object to be compared</param>
        /// <returns>Boolean</returns>
        public override bool Equals(object input)
        {
            input = input is AnyOf anyOf ? anyOf.Obj : input;
            return this.Equals(input as SimulationControl);
        }


        /// <summary>
        /// Returns true if SimulationControl instances are equal
        /// </summary>
        /// <param name="input">Instance of SimulationControl to be compared</param>
        /// <returns>Boolean</returns>
        public bool Equals(SimulationControl input)
        {
            if (input == null)
                return false;
            return base.Equals(input) && 
                    Extension.Equals(this.DoZoneSizing, input.DoZoneSizing) && 
                    Extension.Equals(this.DoSystemSizing, input.DoSystemSizing) && 
                    Extension.Equals(this.DoPlantSizing, input.DoPlantSizing) && 
                    Extension.Equals(this.RunForRunPeriods, input.RunForRunPeriods) && 
                    Extension.Equals(this.RunForSizingPeriods, input.RunForSizingPeriods);
        }


        /// <summary>
        /// Gets the hash code
        /// </summary>
        /// <returns>Hash code</returns>
        public override int GetHashCode()
        {
            unchecked // Overflow is fine, just wrap
            {
                int hashCode = base.GetHashCode();
                if (this.DoZoneSizing != null)
                    hashCode = hashCode * 59 + this.DoZoneSizing.GetHashCode();
                if (this.DoSystemSizing != null)
                    hashCode = hashCode * 59 + this.DoSystemSizing.GetHashCode();
                if (this.DoPlantSizing != null)
                    hashCode = hashCode * 59 + this.DoPlantSizing.GetHashCode();
                if (this.RunForRunPeriods != null)
                    hashCode = hashCode * 59 + this.RunForRunPeriods.GetHashCode();
                if (this.RunForSizingPeriods != null)
                    hashCode = hashCode * 59 + this.RunForSizingPeriods.GetHashCode();
                return hashCode;
            }
        }


    }
}
